### ISSUE TITLE
[BUG][mthreads] Release wheel ships no local llc; the fallback system llc cannot select a 128-bit bf16 store-back

### ISSUE THREAD

**Issue by cheersluvs** [FlagTree/flagtree#997]
## Summary

The `0.6.0+triton3.6` release wheel contains no `triton/backends/mthreads/bin/`
directory, so `compiler.py` falls back to `Path(toolchain_path) / "llc"` — the
`llc` shipped with the MUSA toolkit. That `llc` cannot select a 128-bit bf16
vector load when a kernel stores back to the addresses it read, so a Triton kernel
using that ordinary pattern **cannot be compiled at all** on a release install.

A build from `main` bundles its own `llc` and compiles the same kernel from the
same IR without complaint.

We are not reporting a codegen bug in FlagTree — the IR FlagTree emits is
byte-identical between the two builds. The ask is about what the release ships.

## Environment

| component | version |
|---|---|
| card | MTT S5000 |
| MUSA toolkit / mcc | 4.3.5 (`release_musa_4.3.0`) |
| torch / torch_musa | 2.9.0 / 2.9.0 |
| FlagTree, failing | `0.6.0rc3+mthreads3.6` and `0.6.0+mthreads3.6` (release wheels) |
| FlagTree, working | `0.6.0+mthreads.gitc64a4918` (built from `main`) |

Both cases were run on the same physical card, and for the `PYTHONPATH` test below,
in the same container and the same process image — only the FlagTree package differs.

## Minimal reproduction

Self-contained; no third-party library. The second kernel is a control with the
same width and access pattern, writing to a *different* buffer.

```python
import torch
import triton
import triton.language as tl
import torch_musa  # noqa: F401

@triton.jit
def storeback(p, N: tl.constexpr):
    pid = tl.program_id(0).to(tl.int64)
    offs = pid * N + tl.arange(0, N)
    x = tl.load(p + offs)
    tl.store(p + offs, x * 2)          # back to the addresses just read

@triton.jit
def elsewhere(src, dst, N: tl.constexpr):
    pid = tl.program_id(0).to(tl.int64)
    offs = pid * N + tl.arange(0, N)
    x = tl.load(src + offs)
    tl.store(dst + offs, x * 2)        # control: a different buffer

N, TOK = 512, 64
a = torch.randn(TOK, N, dtype=torch.bfloat16, device="musa")
b = torch.empty_like(a)

for label, fn, args in (("store-back", storeback, (a,)),
                        ("elsewhere ", elsewhere, (a, b))):
    for w in (1, 2, 4, 8):
        try:
            fn[(TOK,)](*args, N=N, num_warps=w, num_stages=2)
            torch.musa.synchronize()
            print(f"  {label}  num_warps={w}  OK")
        except Exception as e:
            print(f"  {label}  num_warps={w}  FAIL  "
                  f"{str(e).strip().splitlines()[0][:80]}")
```

On a **release** install:

```
  store-back  num_warps=1  FAIL  `llc` failed with error code -6
  store-back  num_warps=2  FAIL  `llc` failed with error code -6
  store-back  num_warps=4  OK
  store-back  num_warps=8  OK
  elsewhere   num_warps=1  OK          <- all four pass
  ...
```

On a build from `main`, all eight pass.

The control passing at every `num_warps` is what narrows this to the store-back
aliasing rather than to the 128-bit load width on its own. `num_warps=1` puts 16
bf16 per lane on a 32-lane warp, which lowers to a 128-bit load; `num_warps=4`
drops it to 4 elements per lane and selects fine.

## The decisive test: same IR, two `llc` binaries

The `.llir` FlagTree emits is **byte-identical** between the two builds, and the
`.ttgir` diff is empty — including the intrinsic call that later fails to select:

```llvm
%13 = tail call <8 x bfloat> @llvm.musa.lsu.ld.cache.hint.f.v8bf16.p1(
          ptr addrspace(1) %11, i32 0, i32 2, i32 1, i32 1)
```

Taking the exact IR file and command line that `compiler.py` invokes, and running
it through each `llc`:

```
llc kernel.ll -march=mtgpu -mcpu=mp_31 --opaque-pointers -O2 \
    -filetype=obj -mtgpu-enable-const-calc=1 -o kernel.o
```

| `llc` | md5 | result |
|---|---|---|
| `/usr/local/musa/bin/llc` (MUSA toolkit 4.3.5) | `7d0a51d8…` | aborts, exit 134 |
| `triton/backends/mthreads/bin/llc` (bundled by `main`) | `cec9ff66…` | emits a 5248-byte object |

The failing one prints:

```
LLVM ERROR: Cannot select: v8bf16,ch = MTGPUISD::LSU_LD_CACHE_HINT<(load (s128), addrspace 1)> ...
In function: storeback
...
2.      Running pass 'MTGPU DAG->DAG Pattern Instruction Selection' on function '@storeback'
#7  llvm::SelectionDAGISel::CannotYetSelect(llvm::SDNode*)
```

**Both binaries self-report `LLVM version 14.0.0`**, so a version string cannot
distinguish them — only the md5 can.

## Why the release is affected

`setup_helper.py` fetches `mthreads_local_binary_v0.6.0.tar.gz` and copies `llc`
and `ld.lld` into `third_party/mthreads/bin` (added in #698, merged 2026-06-13).
`compiler.py` uses that when present and otherwise falls back:

```python
if not llc_path and toolchain_path:
    llc_path = str(Path(toolchain_path) / "llc")
```

Observed state, which is what we can attest to:

- the installed release wheel has **no** `triton/backends/mthreads/bin/` at all,
  so the fallback is taken;
- a build from `main` has that directory, with the `llc` whose md5 is above.

We have not determined why the release wheel lacks it. One candidate is that the
release (2026-06-24) predates **#839, "Fix invalid local binary path"** (merged
2026-07-23) — but we have not verified that, and it may be a build-configuration
matter instead.

## Impact

On a release install this is not a slowdown, it is a hard build failure: any
kernel that reads a 512-wide bf16 block and writes it back — an in-place
normalise-and-rotate, for instance — cannot be compiled below `num_warps=4`.

Working around it by forcing `num_warps=4` costs about **2.6x**: on the fixed
`llc` the same operator reaches 1148.6 GB/s against a measured 1332 GB/s ceiling
(86.3%), while the `num_warps=4` workaround measures 447 GB/s (33.5%).

The distortion is worse than the absolute loss. With the defect present, this
operator and vLLM's own portable Triton implementation of the same computation are
*both* forced above their optimum and land at 447 and 446 GB/s — apparent parity.
On the fixed `llc`, at each implementation's own optimum, they are 1148.6 and
982.1. A comparison made on a release install therefore reports "equal" for
implementations that differ by 14–21%.

Also note the failure is easy to misread: the Python-level exception says only
``` `llc` failed with error code -6 ```. -6 is SIGABRT; the `Cannot select` text
goes to the `llc` subprocess's stderr and never reaches the exception message, so
a harness that shows only the exception string makes a real hit look like an
unrelated toolchain glitch. One reviewer on our side concluded "cannot reproduce"
for exactly this reason.

## What we would like

1. A release that ships the working `backends/mthreads/bin/llc`, so that
   `backends.yaml`-pinned installs are not silently routed to a defective
   compiler; or
2. if the bundled binary is intentionally not part of releases, a documented
   minimum MUSA toolkit version for the mthreads backend, plus a clear error when
   the fallback `llc` is used.

Either would let downstream projects state a supported configuration. At the
moment a release install of FlagTree cannot build this operator on S5000 at all,
and there is no version number we can point users at.

## Related

- #982 — a separate mthreads codegen defect reported on the same pinned build.
- We could not find a public issue tracker for the MUSA toolkit itself, which is
  where the `llc` defect ultimately lives; we are happy to re-file there if one
  exists.

**Comment by zhzhcookie** [comment]
`flagtree===0.6.1+mthreads3.6` whl includes `llc`.
